Add save_deg functionality without fenicstools

dbruneau-mie opened the pull request:
This functionality eliminates the use of fenicstools for saving intermediate nodes for higher-order elements, instead making use of PETScDMCollection to create a transfer matrix between higher and lower order (visualization) function spaces.

aslakbergersen commented on turtleFSI/problems/__init__.py:
```diff
@@ -222,21 +220,40 @@ def save_files_visualization(visualization_folder, dvp_, t, save_deg, mesh, **na
+
+        # Interpolate by creating a transfer matrix between higher degree and lower degree visualization function spaces
+        dv_trans = PETScDMCollection.create_transfer_matrix(namespace["FSdv"],namespace["FSdv_viz"])
+        p_trans = PETScDMCollection.create_transfer_matrix(namespace["FSp"],namespace["FSp_viz"])
```
This looks a bit ineffective. I would have initiated `d_viz`, `v_viz`, `p_viz", `dv_trans`, `p_trans` previously in the function, like was done with the fenicstools solution, or is the transver_matrix dependent on the mesh movement? I would still have saved the functions.

aslakbergersen commented on turtleFSI/problems/__init__.py:
```diff
@@ -222,21 +220,40 @@ def save_files_visualization(visualization_folder, dvp_, t, save_deg, mesh, **na
+        # Write results
+        namespace["d_file"].write(d_viz, t)
+        namespace["v_file"].write(v_viz, t)
+        namespace["p_file"].write(p_viz, t)
```
Please try to avoid duplicate code, before it was avoided by having `d_viz` named as `d`, but I see that is difficult there, Instead, I would have created a function which does:
```
def write_solution(d, v, p, d_file, v_file, p_file, t):
    # Name functions
    d.rename("Displacement", "d")
    v.rename("Velocity", "v")
    p.rename("Pressure", "p") 

    # Write results
    d_file.write(d, t)
    v_file.write(v, t)
    p_file.write(p, t)
```

aslakbergersen reviewed (commented):
Looks good, but some small code comments.

KristianValen-Sendstad commented:
Hi guys,

Thanks for the efforts and updates. Aslak, what is the next step?

KVS

On Fri, 21 May 2021 at 12:32, Aslak Bergersen ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Looks good, but some small code comments.
> ------------------------------
>
> In turtleFSI/problems/__init__.py
> <https://github.com/KVSlab/turtleFSI/pull/4#discussion_r636811621>:
>
> > -    p.rename("Pressure", "p")
> -
> -    # Write results
> -    namespace["d_file"].write(d, t)
> -    namespace["v_file"].write(v, t)
> -    namespace["p_file"].write(p, t)
> +    if save_deg > 1: # To save higher-order nodes:
> +
> +        # Create lower-order function for visualization on refined mesh
> +        d_viz = Function(namespace["FSdv_viz"])
> +        v_viz = Function(namespace["FSdv_viz"])
> +        p_viz = Function(namespace["FSp_viz"])
> +
> +        # Interpolate by creating a transfer matrix between higher degree and lower degree visualization function spaces
> +        dv_trans = PETScDMCollection.create_transfer_matrix(namespace["FSdv"],namespace["FSdv_viz"])
> +        p_trans = PETScDMCollection.create_transfer_matrix(namespace["FSp"],namespace["FSp_viz"])
>
> This looks a bit ineffective. I would have initiated d_viz, v_viz, p_viz",
> dv_trans, p_trans` previously in the function, like was done with the
> fenicstools solution, or is the transver_matrix dependent on the mesh
> movement? I would still have saved the functions.
> ------------------------------
>
> In turtleFSI/problems/__init__.py
> <https://github.com/KVSlab/turtleFSI/pull/4#discussion_r636813512>:
>
> > +        # Interpolate by creating a transfer matrix between higher degree and lower degree visualization function spaces
> +        dv_trans = PETScDMCollection.create_transfer_matrix(namespace["FSdv"],namespace["FSdv_viz"])
> +        p_trans = PETScDMCollection.create_transfer_matrix(namespace["FSp"],namespace["FSp_viz"])
> +
> +        d_viz.vector()[:] = dv_trans*d.vector()
> +        v_viz.vector()[:] = dv_trans*v.vector()
> +        p_viz.vector()[:] = p_trans*p.vector()
> +
> +        # Name function
> +        d_viz.rename("Displacement", "d_viz")
> +        v_viz.rename("Velocity", "v_viz")
> +        p_viz.rename("Pressure", "p_viz")
> +        # Write results
> +        namespace["d_file"].write(d_viz, t)
> +        namespace["v_file"].write(v_viz, t)
> +        namespace["p_file"].write(p_viz, t)
>
> Please try to avoid duplicate code, before it was avoided by having d_viz
> named as d, but I see that is difficult there, Instead, I would have
> created a function which does:
>
> def write_solution(d, v, p, d_file, v_file, p_file, t):
>     # Name functions
>     d.rename("Displacement", "d")
>     v.rename("Velocity", "v")
>     p.rename("Pressure", "p")
>
>     # Write results
>     d_file.write(d, t)
>     v_file.write(v, t)
>     p_file.write(p, t)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KVSlab/turtleFSI/pull/4#pullrequestreview-665318791>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4X54ISCPDYI25IF52JHH3TOYZCVANCNFSM45IDY5FQ>
> .
>

aslakbergersen commented:
The next step would be to resolve my comments, which can happen in one of three ways:
1) My comments were clear and you agree, so you make a new commit to the same branch which addresses the comments.
2) My comments were unclear so you ask me what I intended to say and we come up with a way to resolve after a discussion.
3) You disagree with my comments and you express why, and we consider these comments as resolved.

After one of the three happens, I would merge the pull request 👍  Did that answer your question @KristianValen-Sendstad?

dbruneau-mie commented:
Hey Ashlak/KVS, I agree with the comments, I will address them and make
another commit today at some point.

David

On Sat, May 22, 2021, 11:17 AM Aslak Bergersen, ***@***.***>
wrote:

> The next step would be to resolve my comments, which can happen in one of
> three ways:
>
>    1. My comments were clear and you agree, so you make a new commit to
>    the same branch which addresses the comments.
>    2. My comments were unclear so you ask me what I intended to say and
>    we come up with a way to resolve after a discussion.
>    3. You disagree with my comments and you express why, and we consider
>    these comments as resolved.
>
> After one of the three happens, I would merge the pull request 👍 Did
> that answer your question @KristianValen-Sendstad
> <https://github.com/KristianValen-Sendstad>?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KVSlab/turtleFSI/pull/4#issuecomment-846422515>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANWLGEJJXZVMF2MDZHXC36TTO7DINANCNFSM45IDY5FQ>
> .
>

dbruneau-mie commented:
I just updated the repository with the edits to the code, not sure if I need to submit any other pull requests?

aslakbergersen commented:
@dbruneau-mie No need to create another pull request, but you may have followed the git instructions a bit to the letter. I probably did `git add .`, because I had a clean repository with only one change. However, you have added 120 files which should be here, so you would have to clean that up first. Click the "Files changed" to see a list.
